src/engine/SpatialJoin.cpp

```diff
@@ -0,0 +1,295 @@
+#include "SpatialJoin.h"
+#include "util/AllocatorWithLimit.h"
+#include "util/MemorySize/MemorySize.h"
+#include "global/ValueId.h"
+#include "ValuesForTesting.h"
+#include "parser/ParsedQuery.h"
+#include "VariableToColumnMap.h"
+#include "engine/ExportQueryExecutionTrees.h"
+#include "util/GeoSparqlHelpers.h"
+
+SpatialJoin::SpatialJoin(QueryExecutionContext* qec, SparqlTriple triple,
+  std::optional<std::shared_ptr<QueryExecutionTree>> childLeft,
+  std::optional<std::shared_ptr<QueryExecutionTree>> childRight,
+  int maxDist)
+            : Operation(qec) {
+  this->triple = triple;
+  this->maxDist = maxDist;
+  if (childLeft) {
+    this->childLeft = childLeft.value();
+  }
+  if (childRight) {
+    this->childRight = childRight.value();
+  }
+  
+  if (triple._s.isVariable() && triple._o.isVariable()) {
+    leftChildVariable = triple._s.getVariable();
+    rightChildVariable = triple._o.getVariable();
+  } else {
+    AD_THROW("SpatialJoin needs two variables");
+  }
+}
+
+std::shared_ptr<SpatialJoin> SpatialJoin::addChild(
+      std::shared_ptr<QueryExecutionTree> child, Variable varOfChild){
+  if (varOfChild == leftChildVariable) {
+    childLeft = child;
```

Comment:
This function must definitely be `const` otherwise it breaks the query planning. Most notabley you must not overwrite the children of the current instance (you return a new one, that's where you put the changes exclusively).